### PR TITLE
Add props to tell whether locks have been taken immediately

### LINKED ISSUE #111: Add something to IDistributedAppLock that communicates whether we waited or not
```csharp
var myThing = loadMyThingBasedOnSomeThing(x => x.SomeProp == "abc");

using (var theLock = await appLockProvider.Acquire($"lockname-{myThing.Id}"))
{
	// we have no easy way of knowing whether we waited or not,
	// so we will have to reload our thing
	myThing = loadMyThingById(myThing.Id);

	...
}
```

If the `theLock` could tell us whether we waited or not, then we could optimize the above. E.g.

```csharp
var myThing = loadMyThingBasedOnSomeThing(x => x.SomeProp == "abc");

using (var theLock = await appLockProvider.Acquire($"lockname-{myThing.Id}"))
{
	if (!theLock.WasAcquiredInstantly)
	{
		// we waited for the lock, so we probably
		// need to reload our thing
		myThing = loadMyThingById(myThing.Id);
	}
	
	...
}
```

### PR DESCRIPTION
**Pull request by nover**
This PR:

- **adds** prop `WasAcquiredInstantly` that reflects whether the lock was acquired without invoking retry logic
- **adds** prop `TimeUsedToAcquire` that reflects the total amount of time spent acquiring the lock

Breaking change if someone out there has decided to extend our interfaces!

### CODE REVIEW

**Comment by pingvinen** [comment]
At first glance it looks alright, but some tests should help prove it :)

File: src/redis/test-functional/Locking/RedisDistributedAppLockProviderTest.cs [hunk 1]
```diff
@@ -100,6 +100,8 @@ public void Test_acquire_lock_with_timeout_works()
                 Task.Delay(TimeSpan.FromMilliseconds(500));
                 Assert.Equal(lockName, secondLock.Name);
                 Assert.True(secondLock.IsActive);
+                Assert.False((secondLock.WasAcquiredInstantly));
```
Comment (pingvinen) [inline src/redis/test-functional/Locking/RedisDistributedAppLockProviderTest.cs:104]: **issue**: we need to assert that this is true on the `firstLock`, right?

File: src/sqlserver/test-functional/Locking/SqlServerDistributedAppLockProviderTests.cs [hunk 2]
```diff
@@ -139,6 +139,8 @@ public void Test_acquire_lock_with_timeout_works()
                 Task.Delay(TimeSpan.FromMilliseconds(500));
                 Assert.Equal(lockName, secondLock.Name);
                 Assert.True(secondLock.IsActive);
+                Assert.False(secondLock.WasAcquiredInstantly);
```
Comment (pingvinen) [inline src/sqlserver/test-functional/Locking/SqlServerDistributedAppLockProviderTests.cs:143]: **issue**: assert on firstLock as well

File: src/test-unit/Redis/Locking/RedisDistributedAppLockTest.cs [hunk 3]
```diff
@@ -179,5 +179,15 @@ public async Task Properly_traps_other_exceptions()
...
         }
+        
+        [Fact]
+        public async Task Does_set_lock_take_time_and_happened_immediately()
+        {
+            var handle = new RedisDistributedAppLock(_manager, _rng);
+            await handle.AcquireLockAsync(_defaultLockName);
+            
+            Assert.True(handle.WasAcquiredInstantly);
+            Assert.True(handle.TimeUsedToAcquire.TotalMilliseconds > 0);
```
Comment (pingvinen) [inline src/test-unit/Redis/Locking/RedisDistributedAppLockTest.cs]: **suggestion**: Asserting on `Ticks` might be more robust, as we could get lucky with a test that manages to connect faster than 1 ms thereby creating a false negative.

**Review by pingvinen — CHANGES_REQUESTED** [review]

**Review by pingvinen — APPROVED** [review]

**Comment by pingvinen** [comment]
Ah damn... we forgot to update the documentation... sigh
